### ISSUE TITLE
Text and Table on the same page issue

### ISSUE THREAD

**Issue by Tendulkarx** [dhorions/boxable#89]
Hi All, I am hoping I can get some help or direction as to what to do about the following issues.

I need to produce a one page document that comprises of an Image, a table and text all on the same page.

I can produce all two of those on two separate pages at the moment and have spent near three weeks to try to work this out.

When I the table is loaded on it own using BaseTable baseTable = new BaseTable() all works fine. 

I then try to add some text below the table either at a specific location or just at the bottom using 
 PDPageContentStream disclaimerContent = new PDPageContentStream(doc, page), but instead of adding my text to the bottom of the page or below the table, It removes the table and places the text in it place.

I don't seem to be getting anywhere with it at all and I almost sure I have checked every last page on the internet to do with Boxable to find an example but nothing comes up.

Can anyone shed some light on this for me? perhaps show an example ?

**Comment by flurinBoonea** [comment]
Below a sample of a page of our software which does exactly this.

Render a PDF page with text, image and a table.

(some boring details deleted, but you should get the idea)

Best

Flurin


Overall page:
public void renderPages() throws Exception {


    PDPage page = getDefaultProvder().nextPage();

    PDPageContentStream contentStream = new PDPageContentStream(getMainDocument(), page);
    setTitle(contentStream);
    setOverviewTable(page);
    setOverviewImage(contentStream);
    contentStream.close();


}

private void setTitle(PDPageContentStream contentStream) throws Exception {
    contentStream.beginText();
    contentStream.setFont(getTitleFont(), 22);
    contentStream.moveTextPositionByAmount(50, 700);
    contentStream.drawString(getAccountPage().getCompanyURL() + " Lead Overview");
    contentStream.endText();
}


 private void setOverviewTable(PDPage page) throws IOException {

        float margin = 50;
        PDDocument doc = getMainDocument();

        float yStartNewPage = page.getMediaBox().getHeight() - (2 * margin);
        float tableWidth = page.getMediaBox().getWidth() - (2 * margin);

        boolean drawContent = true;
        float yStart = yStartNewPage;
        float bottomMargin = 70;
        BaseTable table = new BaseTable(550, yStartNewPage, bottomMargin, tableWidth, margin, doc, page, true, drawContent);

       	…………..
        cell.setFont(getTextFont());
        table.draw();


    }


private void setOverviewImage(PDPageContentStream contentStream) throws Exception {


        PDImageXObject pdImage = PDImageXObject.createFromFile("chart.png", getMainDocument());
        int w = pdImage.getWidth();
        int h = pdImage.getHeight();
        int newh = 100;
        double ration = (double) h / (double) newh;
        int neww = (int) (w / ration);
        contentStream.drawImage(pdImage, 50, 25, 500, 300);

    }




 					
Flurin Capaul		•		CEO			
Boonea AG		•		The future of sales technology			
t. +41 79 634 00 51		|		e. flurin@boonea.com					
w. http://boonea.com		|							
s. Sihlquai 131					
p. CH-8005 Zurich		|		p. Switzerland					
						


Am 3. Februar 2017 um 16:29:46, Tendulkarx (notifications@github.com) schrieb:

Hi All, I am hoping I can get some help or direction as to what to do about the following issues.

I need to produce a one page document that comprises of an Image, a table and text all on the same page.

I can produce all two of those on two separate pages at the moment and have spent near three weeks to try to work this out.

When I the table is loaded on it own using BaseTable baseTable = new BaseTable() all works fine.

I then try to add some text below the table either at a specific location or just at the bottom using
PDPageContentStream disclaimerContent = new PDPageContentStream(doc, page), but instead of adding my text to the bottom of the page or below the table, It removes the table and places the text in it place.

I don't seem to be getting anywhere with it at all and I almost sure I have checked every last page on the internet to do with Boxable to find an example but nothing comes up.

Can anyone shed some light on this for me? perhaps show an example ?

**Comment by Tendulkarx** [comment]
Hi @flurinBoonea ,

Thanks a million for the reply with the example. I have two small questions,
how did you set up getMainDocument() and getDefaultProvder()

I am guessing thats where the PDDcoument is setup?

**Comment by Tendulkarx** [comment]
Hi @flurinBoonea,
Don't mean to be a pest but would really like your help to answer the above questions?

I still manage to achieve the result, but I suspect is down to how the two methods getMainDocument() and getDefaultProvder() are set up.

Any input would be highly appreciated either by you or anyone else.
Many thanks
O'Neil

**Comment by flurinBoonea** [comment]
I hope this helps!

The provider is a boxable PageProvider - initalized as follows (standard according to doc)

provider = new DefaultPageProvider(
       mainDocument,
       new PDRectangle(
               PDRectangle.A4.getWidth(),
               PDRectangle.A4.getHeight()
       ));


This returns a PDDocument which you setup before according to the documentation.

public PDDocument getMainDocument(){
    return mainDocument;
}
 					
Flurin Capaul		•		CEO			
Boonea AG		•		The future of sales technology			
t. +41 79 634 00 51		|		e. flurin@boonea.com					
w. http://boonea.com		|							
s. Sihlquai 131					
p. CH-8005 Zurich		|		p. Switzerland					
						


Am 6. Februar 2017 um 10:30:41, Tendulkarx (notifications@github.com) schrieb:

Hi @flurinBoonea,
Don't mean to be a pest but would really like your help to answer the above questions?

I still manage to achieve the result, but I suspect is down to how the two methods getMainDocument() and getDefaultProvder() are set up.

Any input would be highly appreciated either by you or anyone else.
Many thanks
O'Neil

**Comment by Tendulkarx** [comment]
Hi @flurinBoonea 

Thanks again, I will try this now. you refer to a documentation which I have been looking for. do you mean looking at the classes or there is an actual documentation somewhere. I have looked and haven't seen any myself.

**Comment by flurinBoonea** [comment]
Fair point! 

There’s no real documentation agreed - I’m unsure how I did it but I guess I just read through various Java examples directly on the Git Hub page.
eg stuff like this:

https://github.com/dhorions/boxable/blob/master/src/test/java/be/quodlibet/boxable/DataTableTest.java




 					
Flurin Capaul		•		CEO			
Boonea AG		•		The future of sales technology			
t. +41 79 634 00 51		|		e. flurin@boonea.com					
w. http://boonea.com		|							
s. Sihlquai 131					
p. CH-8005 Zurich		|		p. Switzerland					
						


Am 6. Februar 2017 um 10:49:23, Tendulkarx (notifications@github.com) schrieb:

Hi @flurinBoonea

Thanks again, I will try this now. you refer to a documentation which I have been looking for. do you mean looking at the classes or there is an actual documentation somewhere. I have looked and haven't seen any myself.

**Comment by Tendulkarx** [comment]
I would kill for a documentation on this trust me. but seriously, Thank you very much for the help. I am spending the day today trying this and if I don't manage to get it then I think I will look into other ways as I have spent far too long on trying to achieve this I think. 

I appreciate your help none the less.
O'Neil

**Comment by flurinBoonea** [comment]
I wrote a small sample, of rendering a title, adding a picture and adding a table.
You need Apache PDFBox and boxable for this to run - also replace the “"./resources/logo-boonea-rgb-highres.png” with a path to a picture you have lying around.

Let me know of this helps.

Regards

Flurin

public class SimpleExample {

    public static void main(String args[]) throws Exception {

        PDPage myPage = new PDPage(PDRectangle.A4);
        PDDocument mainDocument = new PDDocument();


        PDPageContentStream contentStream = new PDPageContentStream(mainDocument, myPage);


        //setTitle
        contentStream.beginText();
        contentStream.setFont(PDType1Font.HELVETICA_BOLD, 22);
        contentStream.moveTextPositionByAmount(50, 700);
        contentStream.drawString("This is a title");
        contentStream.endText();


        //setImage

        PDImageXObject pdImage = PDImageXObject.createFromFile("./resources/logo-boonea-rgb-highres.png", mainDocument);

        int w = pdImage.getWidth();
        int h = pdImage.getHeight();

        if (h > 100) {
            int newh = 100;
            double ration = (double) h / (double) newh;
            int neww = (int) (w / ration);
            contentStream.drawImage(pdImage, 50, 575, neww, newh);
        } else {
            contentStream.drawImage(pdImage, 50, 575, w, h);
        }

        //Dummy Table
        float margin = 50;
        float yStartNewPage = myPage.getMediaBox().getHeight() - (2 * margin);
        float tableWidth = myPage.getMediaBox().getWidth() - (2 * margin);

        boolean drawContent = true;
        float yStart = yStartNewPage;
        float bottomMargin = 70;
        BaseTable table = new BaseTable(550, yStartNewPage, bottomMargin, tableWidth, margin, mainDocument, myPage, true, drawContent);


        Row<PDPage> headerRow = table.createRow(15f);
        Cell<PDPage> cell = headerRow.createCell(100, "Header");
        table.addHeaderRow(headerRow);


        Row<PDPage> row = table.createRow(12);
        cell = row.createCell(30, "Data 1");
        cell = row.createCell(70, "Some value");

        table.draw();


        contentStream.close();
        mainDocument.addPage(myPage);
        mainDocument.save("testfile.pdf");
        mainDocument.close();
    }
}

 					
Flurin Capaul		•		CEO			
Boonea AG		•		The future of sales technology			
t. +41 79 634 00 51		|		e. flurin@boonea.com					
w. http://boonea.com		|							
s. Sihlquai 131					
p. CH-8005 Zurich		|		p. Switzerland					
						


Am 6. Februar 2017 um 10:57:17, Tendulkarx (notifications@github.com) schrieb:

I would kill for a documentation on this trust me. but seriously, Thank you very much for the help. I am spending the day today trying this and if I don't manage to get it then I think I will look into other ways as I have spent far too long on trying to achieve this I think.

I appreciate your help none the less.
O'Neil

**Comment by flurinBoonea** [comment]
Here's a small, nicely formatted sample - make sure you replace the filename with something on your disk in order for it to run:

`public class SimpleExample {

    public static void main(String args[]) throws Exception {

        PDPage myPage = new PDPage(PDRectangle.A4);
        PDDocument mainDocument = new PDDocument();


        PDPageContentStream contentStream = new PDPageContentStream(mainDocument, myPage);


        //setTitle
        contentStream.beginText();
        contentStream.setFont(PDType1Font.HELVETICA_BOLD, 22);
        contentStream.moveTextPositionByAmount(50, 700);
        contentStream.drawString("This is a title");
        contentStream.endText();


        //setImage

        PDImageXObject pdImage = PDImageXObject.createFromFile(<<INSERT NAME OF FILE eg PNG  >>, mainDocument);

        int w = pdImage.getWidth();
        int h = pdImage.getHeight();

        if (h > 100) {
            int newh = 100;
            double ration = (double) h / (double) newh;
            int neww = (int) (w / ration);
            contentStream.drawImage(pdImage, 50, 575, neww, newh);
        } else {
            contentStream.drawImage(pdImage, 50, 575, w, h);
        }

        //Dummy Table
        float margin = 50;
        float yStartNewPage = myPage.getMediaBox().getHeight() - (2 * margin);
        float tableWidth = myPage.getMediaBox().getWidth() - (2 * margin);

        boolean drawContent = true;
        float yStart = yStartNewPage;
        float bottomMargin = 70;
        BaseTable table = new BaseTable(550, yStartNewPage, bottomMargin, tableWidth, margin, mainDocument, myPage, true, drawContent);


        Row<PDPage> headerRow = table.createRow(15f);
        Cell<PDPage> cell = headerRow.createCell(100, "Header");
        table.addHeaderRow(headerRow);


        Row<PDPage> row = table.createRow(12);
        cell = row.createCell(30, "Data 1");
        cell = row.createCell(70, "Some value");

        table.draw();


        contentStream.close();
        mainDocument.addPage(myPage);
        mainDocument.save("testfile.pdf");
        mainDocument.close();
    }
}`

**Comment by flurinBoonea** [comment]
I deleted my earlier comments to clean this up.

**Comment by Frulenzo** [comment]
First, sorry for late response. Secondly, @flurinBoonea gave you pretty good example how can you make your scenario but I would add this for image creation. You can always use 'Image' class in next manner :

```java
Image image = new Image(final BufferedImage image)
```
Useful methods for drawing and image scalation:
```java
draw(final PDDocument doc, final PDPageContentStream stream, float x, float y)
scaleByWidth(float width)
scaleByHeight(float height)
scale(float boundWidth, float boundHeight)
```

So at the end you have something like:
```java
Image image = new Image(ImageIO.read(FILE_PATH));
if (image != null) {
   image.draw(document, cos, xStart, yStart);
}
```

**Comment by Tendulkarx** [comment]
@flurinBoonea  You have been such a life saver and I highly appreciated the help. I did manage to get it on one page now those issues have been sorted. so once again thank you.

@Frulenzo I will give the image thing a go after I have worked out how to change the default table styles. and you are right, I did have issues with scaling but I resized my image outside externally and which has had some impact on the quality but I am focusing on one thing at a time so thanks for adding to the solution @flurinBoonea  posted. 

Thanks to you both.
O'Neil
